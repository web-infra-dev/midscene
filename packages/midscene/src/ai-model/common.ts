import type OpenAI from 'openai';
import type {
  ChatCompletionContentPart,
  ChatCompletionSystemMessageParam,
  ChatCompletionUserMessageParam,
} from 'openai/resources';
import {
  COZE_AI_ACTION_BOT_ID,
  COZE_AI_ASSERT_BOT_ID,
  COZE_EXTRACT_INFO_BOT_ID,
  COZE_INSPECT_ELEMENT_BOT_ID,
  callCozeAi,
  transfromOpenAiArgsToCoze,
  useCozeModel,
} from './coze';
import {
  MIDSCENE_MODEL_TEXT_ONLY,
  callToGetJSONObject,
  useOpenAIModel,
} from './openai';

export type AIArgs = [
  ChatCompletionSystemMessageParam,
  ChatCompletionUserMessageParam,
];

export enum AIActionType {
  ASSERT = 0,
  INSPECT_ELEMENT = 1,
  EXTRACT_DATA = 2,
  PLAN = 3,
}

export async function callAiFn<T>(options: {
  msgs: AIArgs;
  AIActionType: AIActionType;
  useModel?: 'openAI' | 'coze';
}) {
  const { useModel, msgs, AIActionType: AIActionTypeValue } = options;
  if (useOpenAIModel(useModel)) {
    const parseResult = await callToGetJSONObject<T>(msgs, AIActionTypeValue);
    return parseResult;
  }

  if (useCozeModel(useModel)) {
    let botId = '';
    switch (AIActionTypeValue) {
      case AIActionType.ASSERT:
        botId = COZE_AI_ASSERT_BOT_ID;
        break;
      case AIActionType.EXTRACT_DATA:
        botId = COZE_EXTRACT_INFO_BOT_ID;
        break;
      case AIActionType.INSPECT_ELEMENT:
        botId = COZE_INSPECT_ELEMENT_BOT_ID;
        break;
      default:
        botId = COZE_AI_ACTION_BOT_ID;
    }
    const cozeMsg = transfromOpenAiArgsToCoze(msgs[1]);
    const parseResult = await callCozeAi<T>({
      ...cozeMsg,
      botId,
    });
    return parseResult;
  }

  throw Error(
    'Cannot find Coze or OpenAI config. You should set at least one of them.',
  );
}

export function transformUserMessages(msgs: ChatCompletionContentPart[]) {
  const textOnly = Boolean(process.env[MIDSCENE_MODEL_TEXT_ONLY]);
  if (!textOnly) return msgs;

  return msgs.reduce((res, msg) => {
    if (msg.type === 'text') {
      res += msg.text;
    }
    return res;
  }, '');
}
