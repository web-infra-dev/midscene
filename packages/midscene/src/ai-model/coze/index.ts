import assert from 'node:assert';
import type { ChatCompletionUserMessageParam } from 'openai/resources';

export const COZE_INSPECT_ELEMENT_BOT_ID =
  process.env.COZE_INSPECT_ELEMENT_BOT_ID || '';
export const COZE_AI_ACTION_BOT_ID = process.env.COZE_AI_ACTION_BOT_ID || '';
export const COZE_AI_ASSERT_BOT_ID = process.env.COZE_AI_ASSERT_BOT_ID || '';
export const COZE_EXTRACT_INFO_BOT_ID =
  process.env.COZE_EXTRACT_INFO_BOT_ID || '';

export const COZE_BOT_TOKEN = 'COZE_BOT_TOKEN';

export function preferCozeModel(preferVendor?: 'coze' | 'openAI') {
  if (preferVendor && preferVendor !== 'coze') return false;
  return (
    process.env[COZE_BOT_TOKEN] &&
    process.env.COZE_INSPECT_ELEMENT_BOT_ID &&
    process.env.COZE_AI_ACTION_BOT_ID &&
    process.env.COZE_AI_ASSERT_BOT_ID &&
    process.env.COZE_EXTRACT_INFO_BOT_ID
  );
}

export async function callCozeAi<T>(options: {
  query: string;
  imgs: Array<string>;
  botId: string;
}): Promise<T> {
  const { query, imgs, botId } = options;
  const completion = await fetch('https://api.coze.com/open_api/v2/chat', {
    method: 'POST',
    headers: {
      Authorization: `Bearer ${process.env[COZE_BOT_TOKEN]}`,
      'Content-Type': 'application/json',
      Accept: '*/*',
      Host: 'api.coze.com',
      Connection: 'keep-alive',
    },
    body: JSON.stringify({
      conversation_id: '123',
      bot_id: botId,
      user: '29032201862555',
      query,
      meta_data: {
        img: imgs.map((imgPath) => {
          return {
            url: imgPath,
          };
        }),
      },
      stream: false,
    }),
  });
  if (!completion.ok) {
    console.error('CozeAI reponse error', completion);
    throw new Error('Network response was not ok');
  }

  const aiResponse = await completion.json();
  if (aiResponse.code !== 0) {
    console.error('CozeAI error response', aiResponse.msg);
    throw new Error(`CozeAI error response ${aiResponse.msg}`);
  }

  if (!aiResponse?.messages || !aiResponse?.messages[0]?.content) {
    console.error('aiResponse', aiResponse);
    throw new Error('aiResponse is undefined', aiResponse);
  }
  const parseContent = aiResponse?.messages[0]?.content;
  assert(parseContent, 'empty content');
  try {
    return JSON.parse(parseContent);
  } catch (err) {
    console.error("can't parse coze content", aiResponse, err);
    throw Error("can't parse coze content");
  }
}

export function transformOpenAiArgsToCoze(msg: ChatCompletionUserMessageParam) {
  if (msg.role !== 'user') throw Error(`can't transform ${msg} to coze args`);
  // const query = '';
  // const imgs = msg.content
  if (typeof msg.content === 'string') {
    return {
      query: msg.content,
      imgs: [],
    };
  }

  return {
    query: msg.content.reduce((res, next) => {
      if (next.type === 'text') {
        res += `\n${next.text}`;
      }
      return res;
    }, ''),
    imgs: msg.content.reduce(
      (res, next) => {
        if (next.type === 'image_url') {
          res.push(next.image_url.url);
        }
        return res;
      },
      [] as Array<string>,
    ),
  };
}
