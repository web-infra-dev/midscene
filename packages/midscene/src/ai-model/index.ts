export { callToGetJSONObject } from './openai';
export { systemPromptToFindElement } from './prompt/element_inspector';
export { describeUserPage } from './prompt/util';

export type { ChatCompletionMessageParam } from 'openai/resources';

export {
  AiInspectElement,
  AiExtractElementInfo,
  AiAssert,
  transformElementPositionToId,
} from './inspect';

export { plan } from './automation';
export { callAiFn } from './common';
