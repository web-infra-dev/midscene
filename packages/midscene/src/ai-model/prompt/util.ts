import assert from 'node:assert';
import { imageInfoOfBase64 } from '@/image';
import type {
  BaseElement,
  BasicSectionQuery,
  Size,
  UIContext,
  UISection,
} from '@/types';
import type { ResponseFormatJSONSchema } from 'openai/resources';

const characteristic =
  'You are a versatile professional in software UI design and testing. Your outstanding contributions will impact the user experience of billions of users.';

const contextFormatIntro = `
The user will give you a screenshot and the texts on it. There may be some none-English characters (like Chinese) on it, indicating it's an non-English app.`;

const ONE_ELEMENT_LOCATOR_PREFIX = 'LOCATE_ONE_ELEMENT';
const ELEMENTS_LOCATOR_PREFIX = 'LOCATE_ONE_OR_MORE_ELEMENTS';
const SECTION_MATCHER_FLAG = 'SECTION_MATCHER_FLAG/';

export function systemPromptToFindElement(queryPrompt: string, multi: boolean) {
  assert(queryPrompt, 'queryPrompt is required');
  return `
${characteristic}
${contextFormatIntro}

Based on the information you get, find ${
    multi ? 'one or more text elements' : 'ONE text element'
  } on the page.
Here is the description: ${queryPrompt}
  
Return in the following JSON format:
{
  "elements": [ // Leave it an empty array when no element is found
    { 
      "id": "id of the element, like 123", 
    },
    // more ...
  ], 
  errors?: [], // string[], error message if any
}
`;
}

const skillSegment = `skill name: segment_a_web_page 
skill content:
Based on the functions and content of various elements on the page, segment the screenshot into different sections like navigation bar, product list, news area, etc. 
Some general rules for segmentation:
* Each section should NOT overlap with each other.
* Each text should only belong to one section.
* [IMPORTANT] Whether the content visually appears to belong to different sections is a significant factor in segmenting the page.
* Analyze the page in a top-to-bottom and left-to-right order.
* The evidence indicates a separate section, for example 
  - The background color of certain parts of the page changes.
  - A section of a page includes a title.
* Provide the following data for each of the UI section you found.
  {
    "name": "name of the section",
    "description": "briefly summarize the key content or usage of this section.",
    "sectionCharacteristics": "In view of the need to distinguish this section from the surrounding sections, explain the characteristics and how to define boundaries and what precautions to take.",
    "textIds": ["5", "6", "7"], // ids of all text elements in this section
  }
`;

const skillExtractData = `skill name: extract_data_from_UI
related input: DATA_DEMAND
skill content: 
* User will give you some data requirements in DATA_DEMAND. Consider the UI context, follow the user's instructions, and provide comprehensive data accordingly.
* There may be some special commands in DATA_DEMAND, please pay extra attention
  - ${ONE_ELEMENT_LOCATOR_PREFIX} and ${ELEMENTS_LOCATOR_PREFIX}: if you see a description that mentions the keyword ${ONE_ELEMENT_LOCATOR_PREFIX} or ${ELEMENTS_LOCATOR_PREFIX}(e.g. follow ${ONE_ELEMENT_LOCATOR_PREFIX} : i want to find ...), it means user wants to locate a specific element meets the description. Return in this way: prefix + the id / comma-separated ids, for example: ${ONE_ELEMENT_LOCATOR_PREFIX}/1 , ${ELEMENTS_LOCATOR_PREFIX}/1,2,3 . If not found, keep the prefix and leave the suffix empty, like ${ONE_ELEMENT_LOCATOR_PREFIX}/ .`;

export function promptsOfSectionQuery(
  constraints: BasicSectionQuery[],
): string {
  if (!constraints.length) {
    return '';
  }
  const instruction =
    'Use your segment_a_web_page skill to find the following section(s)';
  const singleSection = (c: BasicSectionQuery) => {
    assert(
      c.name || c.description,
      'either `name` or `description` is required to define a section constraint',
    );

    const number = 'One section';
    const name = c.name ? `named \`${c.name}\`` : '';
    const description = c.description
      ? `, usage or criteria : ${c.description}`
      : '';
    const basic = `* ${number} ${name}${description}`;

    return basic;
  };
  return `${instruction}\n${constraints.map(singleSection).join('\n')}`;
}

export function systemPromptToExtract() {
  return `
You are a versatile professional in software UI design and testing. Your outstanding contributions will impact the user experience of billions of users.
The user will give you a screenshot and the texts on it. There may be some none-English characters (like Chinese) on it, indicating it's an non-English app.

You have the following skills:

skill name: extract_data_from_UI
related input: DATA_DEMAND
skill content: 
* User will give you some data requirements in DATA_DEMAND. Consider the UI context, follow the user's instructions, and provide comprehensive data accordingly.
* There may be some special commands in DATA_DEMAND, please pay extra attention
  - LOCATE_ONE_ELEMENT and LOCATE_ONE_OR_MORE_ELEMENTS: if you see a description that mentions the keyword LOCATE_ONE_ELEMENT
  - LOCATE_ONE_OR_MORE_ELEMENTS(e.g. follow LOCATE_ONE_ELEMENT : i want to find ...), it means user wants to locate a specific element meets the description. 

Return in this way: prefix + the id / comma-separated ids, for example: LOCATE_ONE_ELEMENT/1 , LOCATE_ONE_OR_MORE_ELEMENTS/1,2,3 . If not found, keep the prefix and leave the suffix empty, like LOCATE_ONE_ELEMENT/ .



Return in the following JSON format:
{
  language: "en", // "en" or "zh", the language of the page. Use the same language to describe section name, description, and similar fields.
  data: any, // the extracted data from extract_data_from_UI skill. Make sure both the value and scheme meet the DATA_DEMAND.
  errors: [], // string[], error message if any
}
`;
}

export const extractDataSchema: ResponseFormatJSONSchema = {
  type: 'json_schema',
  json_schema: {
    name: 'extract_data',
    strict: true,
    schema: {
      type: 'object',
      properties: {
        language: {
          type: 'string',
          enum: ['en', 'zh'],
          description: 'The language of the page',
        },
        data: {
          type: 'object',
          description: 'The extracted data from extract_data_from_UI skill',
        },
        errors: {
          type: 'array',
          items: {
            type: 'string',
          },
          description: 'Error messages, if any',
        },
      },
      required: ['language', 'data', 'errors'],
      additionalProperties: false,
    },
  },
};

export function systemPromptToAssert() {
  return `
${characteristic}
${contextFormatIntro}

Based on the information you get, Return assertion judgment:

Return in the following JSON format:
{
  thought: string, // string, the thought of the assertion
  pass: true, // true or false, whether the assertion is passed
}
`;
}

export const assertSchema: ResponseFormatJSONSchema = {
  type: 'json_schema',
  json_schema: {
    name: 'assert',
    strict: true,
    schema: {
      type: 'object',
      properties: {
        thought: {
          type: 'string',
          description: 'The thought process behind the assertion',
        },
        pass: {
          type: 'boolean',
          description: 'Whether the assertion passed or failed',
        },
      },
      required: ['thought', 'pass'],
      additionalProperties: false,
    },
  },
};

/*
To modify the response format:
  1. update the function `describeSectionResponseFormat` here
  2. update `expandLiteSection` in insight/utils.ts
  3. update `UISection` and `LiteUISection` in types.ts
*/

export function describeSize(size: Size) {
  return `${size.width} x ${size.height}`;
}

export function describeTextFormat() {
  // use `right` and `bottom` to help AI reduce the feasibility of performing computations
  return `
The following texts elements are formatted in the following way: 
id(string), left, top, right, bottom, content(may be truncated)`;
}

type PromptElementType = {
  id: BaseElement['id'];
  attributes: BaseElement['attributes'];
  rect: BaseElement['rect'];
  content: BaseElement['content'];
};

export function describeElement(
  elements: (Pick<BaseElement, 'rect' | 'content'> & { id: string })[],
) {
  const sliceLength = 80;
  return elements
    .map((item) =>
      [
        item.id,
        item.rect.left,
        item.rect.top,
        item.rect.left + item.rect.width,
        item.rect.top + item.rect.height,
        item.content.length > sliceLength
          ? `${item.content.slice(0, sliceLength)}...`
          : item.content,
      ].join(', '),
    )
    .join('\n');
}

export function describeSectionInputFormat() {
  return `
The sections are formatted in the following way:
  left, top, right, bottom, color-on-the-diagram, section-name
`;
}

export function describeSections(
  sections: UISection[],
  colorOfSectionName: (name: string) => string,
) {
  return sections
    .map((item) =>
      [
        item.rect.left,
        item.rect.top,
        item.rect.left + item.rect.width,
        item.rect.top + item.rect.height,
        colorOfSectionName(item.name),
        item.name,
      ].join(', '),
    )
    .join('\n');
}

export function truncateText(text: string) {
  const maxLength = 50;
  if (text && text.length > maxLength) {
    return `${text.slice(0, maxLength)}...`;
  }
  return text;
}

export async function describeUserPage<
  ElementType extends BaseElement = BaseElement,
>(context: Omit<UIContext<ElementType>, 'describer'>) {
  const { screenshotBase64 } = context;
  let width: number;
  let height: number;

  if (context.size) {
    ({ width, height } = context.size);
  } else {
    const imgSize = await imageInfoOfBase64(screenshotBase64);
    ({ width, height } = imgSize);
  }

  const elementsInfo = context.content;
  const idElementMap: Record<string, ElementType> = {};
  elementsInfo.forEach((item) => {
    idElementMap[item.id] = item;
    return { ...item };
  });

  const elementInfosDescription = cropFieldInformation(elementsInfo);

  return {
    description: `
    {
      // The size of the page
      "pageSize": ${describeSize({ width, height })},\n
      
      // json description of the element
      "elementInfos": ${JSON.stringify(elementInfosDescription)}
    }`,
    elementById(id: string) {
      assert(typeof id !== 'undefined', 'id is required for query');
      const item = idElementMap[`${id}`];
      return item;
    },
  };
}

function cropFieldInformation(elementsInfo: BaseElement[]) {
  const elementInfosDescription: Array<PromptElementType> = elementsInfo.map(
    (item) => {
      const { id, attributes = {}, rect, content } = item;
      const tailorContent = truncateText(content);
      const tailorAttributes = Object.keys(attributes).reduce(
        (res, currentKey: string) => {
          const attributeVal = (attributes as any)[currentKey];
          res[currentKey] = truncateText(attributeVal);
          return res;
        },
        {} as BaseElement['attributes'],
      );

      return {
        id,
        attributes: tailorAttributes,
        rect,
        content: tailorContent,
      };
    },
  );
  return JSON.stringify(elementInfosDescription);
}

/**
 * elements
 */
export function retrieveElement(
  prompt: string,
  opt?: { multi: boolean },
): string {
  if (opt?.multi) {
    return `follow ${ELEMENTS_LOCATOR_PREFIX}: ${prompt}`;
  }
  return `follow ${ONE_ELEMENT_LOCATOR_PREFIX}: ${prompt}`;
}

export function ifElementTypeResponse(response: string): boolean {
  if (typeof response !== 'string') {
    return false;
  }
  return (
    response.startsWith(ONE_ELEMENT_LOCATOR_PREFIX) ||
    response.startsWith(ELEMENTS_LOCATOR_PREFIX)
  );
}

export function splitElementResponse(
  response: string,
): string | null | string[] {
  const oneElementSplitter = `${ONE_ELEMENT_LOCATOR_PREFIX}/`;
  if (response.startsWith(oneElementSplitter)) {
    const id = response.slice(oneElementSplitter.length);
    if (id.indexOf(',') >= 0) {
      console.warn(`unexpected comma in one element response: ${id}`);
    }
    return id ? id : null;
  }

  const elementsSplitter = `${ELEMENTS_LOCATOR_PREFIX}/`;
  if (response.startsWith(elementsSplitter)) {
    const idsString = response.slice(elementsSplitter.length);
    if (!idsString) {
      return [];
    }
    return idsString.split(',');
  }

  return null;
}

/**
 * sections
 */

export function retrieveSection(prompt: string): string {
  return `${SECTION_MATCHER_FLAG}${prompt}`;
}

export function extractSectionQuery(input: string): string | false {
  if (typeof input === 'string' && input.startsWith(SECTION_MATCHER_FLAG)) {
    return input.slice(SECTION_MATCHER_FLAG.length);
  }
  return false;
}
