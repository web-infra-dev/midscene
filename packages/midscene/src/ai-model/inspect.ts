import assert from 'node:assert';
import type {
  AIAssertionResponse,
  AIElementParseResponse,
  AISectionParseResponse,
  BaseElement,
  UIContext,
} from '@/types';
import type {
  ChatCompletionSystemMessageParam,
  ChatCompletionUserMessageParam,
} from 'openai/resources';
import { AIActionType, callAiFn, transformUserMessages } from './common';
import {
  multiDescription,
  systemPromptToFindElement,
} from './prompt/element_inspector';
import {
  describeUserPage,
  systemPromptToAssert,
  systemPromptToExtract,
} from './prompt/util';

export type AIArgs = [
  ChatCompletionSystemMessageParam,
  ChatCompletionUserMessageParam,
];

export async function AiInspectElement<
  ElementType extends BaseElement = BaseElement,
>(options: {
  context: UIContext<ElementType>;
  multi: boolean;
  findElementDescription: string;
  callAI?: typeof callAiFn<AIElementParseResponse>;
  useModel?: 'coze' | 'openAI';
}) {
  const { context, multi, findElementDescription, callAI, useModel } = options;
  const { screenshotBase64 } = context;
  const { description, elementById } = await describeUserPage(context);

  const systemPrompt = systemPromptToFindElement();

  const msgs: AIArgs = [
    { role: 'system', content: systemPrompt },
    {
      role: 'user',
      content: transformUserMessages([
        {
          type: 'image_url',
          image_url: {
            url: screenshotBase64,
          },
        },
        {
          type: 'text',
          text: `
    pageDescription: \n
    ${description}

    Here is the description of the findElement. Just go ahead:
    =====================================
    ${JSON.stringify({
      description: findElementDescription,
      multi: multiDescription(multi),
    })}
    =====================================
  `,
        },
      ]),
    },
  ];

  if (callAI) {
    const parseResult = await callAI({
      msgs,
      AIActionType: AIActionType.INSPECT_ELEMENT,
      useModel,
    });
    return {
      parseResult,
      elementById,
    };
  }
  const inspectElement = await callAiFn<AIElementParseResponse>({
    msgs,
    AIActionType: AIActionType.INSPECT_ELEMENT,
    useModel,
  });

  return {
    parseResult: inspectElement,
    elementById,
  };
}

export async function AiExtractElementInfo<
  T,
  ElementType extends BaseElement = BaseElement,
>(options: {
  dataQuery: string | Record<string, string>;
  context: UIContext<ElementType>;
  useModel?: 'coze' | 'openAI';
}) {
  const { dataQuery, context, useModel } = options;
  const systemPrompt = systemPromptToExtract();

  const { screenshotBase64 } = context;
  const { description, elementById } = await describeUserPage(context);

  const msgs: AIArgs = [
    { role: 'system', content: systemPrompt },
    {
      role: 'user',
      content: transformUserMessages([
        {
          type: 'image_url',
          image_url: {
            url: screenshotBase64,
          },
        },
        {
          type: 'text',
          text: `
pageDescription: ${description}

Use your extract_data_from_UI skill to find the following data, placing it in the \`data\` field
DATA_DEMAND start:
=====================================
${
  typeof dataQuery === 'object'
    ? `return in key-value style object, keys are ${Object.keys(dataQuery).join(',')}`
    : ''
};
${typeof dataQuery === 'string' ? dataQuery : JSON.stringify(dataQuery, null, 2)}
=====================================
DATA_DEMAND ends.
          `,
        },
      ]),
    },
  ];

  const parseResult = await callAiFn<AISectionParseResponse<T>>({
    msgs,
    useModel,
    AIActionType: AIActionType.EXTRACT_DATA,
  });
  return {
    parseResult,
    elementById,
  };
}

export async function AiAssert<
  ElementType extends BaseElement = BaseElement,
>(options: {
  assertion: string;
  context: UIContext<ElementType>;
  useModel?: 'coze' | 'openAI';
}) {
  const { assertion, context, useModel } = options;

  assert(assertion, 'assertion should be a string');

  const { screenshotBase64 } = context;
  const { description, elementById } = await describeUserPage(context);
  const systemPrompt = systemPromptToAssert();

  const msgs: AIArgs = [
    { role: 'system', content: systemPrompt },
    {
      role: 'user',
      content: transformUserMessages([
        {
          type: 'image_url',
          image_url: {
            url: screenshotBase64,
          },
        },
        {
          type: 'text',
          text: `
    pageDescription: \n
    ${description}
    Here is the description of the assertion. Just go ahead:
    =====================================
    ${assertion}
    =====================================
  `,
        },
      ]),
    },
  ];

  const assertResult = await callAiFn<AIAssertionResponse>({
    msgs,
    AIActionType: AIActionType.ASSERT,
    useModel,
  });
  return assertResult;
}
export { callAiFn };
