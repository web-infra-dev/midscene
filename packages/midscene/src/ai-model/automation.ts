import assert from 'node:assert';
import type { AIUsageInfo, PlanningAIResponse, UIContext } from '@/types';
import { PromptTemplate } from '@langchain/core/prompts';
import { AIActionType, type AIArgs, callAiFn } from './common';
import {
  automationUserPrompt,
  systemPromptToTaskPlanning,
  taskBackgroundContext,
} from './prompt/planning';
import { describeUserPage } from './prompt/util';

export async function plan(
  userPrompt: string,
  opts: {
    whatHaveDone?: string;
    originalPrompt?: string;
    context: UIContext;
    callAI?: typeof callAiFn<PlanningAIResponse>;
  },
): Promise<PlanningAIResponse> {
  const { callAI, context } = opts || {};
  const { screenshotBase64, screenshotBase64WithElementMarker } = context;
  const { description: pageDescription, elementByPosition } =
    await describeUserPage(context);

  const systemPrompt = await systemPromptToTaskPlanning();
  const userInstructionPrompt = await automationUserPrompt.format({
    pageDescription,
    userPrompt,
    taskBackgroundContext: taskBackgroundContext(
      opts.originalPrompt,
      opts.whatHaveDone,
    ),
  });

  const msgs: AIArgs = [
    { role: 'system', content: systemPrompt },
    {
      role: 'user',
      content: [
        {
          type: 'image_url',
          image_url: {
            url: screenshotBase64WithElementMarker || screenshotBase64,
            detail: 'high',
          },
        },
        {
          type: 'text',
          text: userInstructionPrompt,
        },
      ],
    },
  ];

  const call = callAI || callAiFn;
  const { content, usage } = await call(msgs, AIActionType.PLAN);
  const planFromAI = content;

  const actions = planFromAI?.actions || [];
  assert(planFromAI, "can't get plans from AI");
  assert(
    actions.length > 0,
    `Failed to plan actions with context: ${planFromAI.error}`,
  );

  return planFromAI;
}
