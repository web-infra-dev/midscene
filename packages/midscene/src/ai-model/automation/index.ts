import assert from 'node:assert';
import type { PlanningAIResponse, PlanningAction, UIContext } from '@/types';
import {
  AIActionType,
  type AIArgs,
  callAiFn,
  transformUserMessages,
} from '../common';
import { describeUserPage } from '../prompt/util';
import { systemPromptToTaskPlanning } from './planning';

export async function plan(
  userPrompt: string,
  opts: {
    context: UIContext;
    callAI?: typeof callAiFn<PlanningAIResponse>;
  },
  useModel?: 'coze' | 'openAI',
): Promise<{ plans: PlanningAction[] }> {
  const { callAI, context } = opts || {};
  const { screenshotBase64 } = context;
  const { description: pageDescription } = await describeUserPage(context);
  let planFromAI: PlanningAIResponse | undefined;

  const systemPrompt = systemPromptToTaskPlanning();
  const msgs: AIArgs = [
    { role: 'system', content: systemPrompt },
    {
      role: 'user',
      content: transformUserMessages([
        {
          type: 'image_url',
          image_url: {
            url: screenshotBase64,
            detail: 'high',
          },
        },
        {
          type: 'text',
          text: `
            pageDescription:\n 
            ${pageDescription}
            \n
            Here is the description of the task. Just go ahead:
            =====================================
            ${userPrompt}
            =====================================
          `,
        },
      ]),
    },
  ];

  if (callAI) {
    planFromAI = await callAI({
      msgs,
      AIActionType: AIActionType.PLAN,
      useModel,
    });
  } else {
    planFromAI = await callAiFn({
      msgs,
      AIActionType: AIActionType.PLAN,
      useModel,
    });
  }

  const actions = planFromAI?.actions || [];

  assert(planFromAI, "can't get planFromAI");
  assert(actions && actions.length > 0, 'no actions in ai plan');

  if (planFromAI.error) {
    throw new Error(planFromAI.error);
  }

  // actions.forEach((task) => {
  //   if (task.type === 'Error') {
  //     throw new Error(task.thought);
  //   }
  // });

  return { plans: actions };
}
