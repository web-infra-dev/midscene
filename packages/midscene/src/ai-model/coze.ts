import type { BaseElement, UIContext } from '@/types';
import fetch from 'node-fetch';
import { multiDescription } from './prompt/element_inspector';
import { describeUserPage } from './prompt/util';

const INSPECT_ELEMENT_BOT_ID = '7390985487806775304';

export async function CozeAiInspectElement<
  ElementType extends BaseElement = BaseElement,
>(msg: {
  findElementDescription: string;
  context: UIContext<ElementType>;
  multi?: boolean;
}) {
  const { multi = false, findElementDescription, context } = msg;
  const { screenshotBase64 } = context;
  const { description, elementById } = await describeUserPage(context);
  const query = JSON.stringify({
    description: findElementDescription,
    multi: multiDescription(multi),
    elementDescription: description,
  });
  const parseResult = await fetch('https://api.coze.com/open_api/v2/chat', {
    method: 'POST',
    headers: {
      Authorization: `Bearer ${process.env.COZE_BOT_TOKEN}`,
      'Content-Type': 'application/json',
      Accept: '*/*',
      Host: 'api.coze.com',
      Connection: 'keep-alive',
    },
    body: JSON.stringify({
      conversation_id: '123',
      bot_id: INSPECT_ELEMENT_BOT_ID,
      user: '29032201862555',
      query,
      meta_data: {
        img: [
          {
            url: screenshotBase64,
          },
        ],
      },
      stream: false,
    }),
  })
    .then((response) => {
      if (!response.ok) {
        console.log('response', response);
        throw new Error('Network response was not ok');
      }
      return response.json();
    })
    .then((data) => {
      const parseContent = data?.messages[0]?.content;
      if (parseContent) {
        return JSON.parse(parseContent);
      }
      return data;
    })
    .catch((error) => {
      console.error('Fetch error:', error);
    });
  return {
    parseResult,
    elementById,
  };
}
