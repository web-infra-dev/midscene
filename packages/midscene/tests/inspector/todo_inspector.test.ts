import { it, expect } from 'vitest';
import path from 'path';
import { getPageTestData, repeat, runTestCases, writeFileSyncWithDir } from './util';
import { AiInspectElement } from '@/ai-model';


const testTodoCases = [
  {
    description: '任务输入框',
    multi: false,
  },
  {
    description: '任务列表中的第二项',
    multi: false,
  },
  {
    description: '第二项任务右边的删除按钮',
    multi: false,
  },
  {
    description: '任务列表中第三项左边的勾选按钮',
    multi: false,
  },
  {
    description: '任务列表下面的 Completed 状态按钮',
    multi: false,
  },
];


repeat(2, (repeatIndex) => {
  it('todo: inspect element', async () => {
    const { context } = await getPageTestData(path.join(__dirname, './test-data/todo'));
  
    const { aiResponse, filterUnStableinf } = await runTestCases(testTodoCases, async (testCase)=>{
        const { parseResult } = await AiInspectElement({
            context,
            multi: testCase.multi,
            findElementDescription: testCase.description,
        });
        return parseResult;
    });
    writeFileSyncWithDir(path.join(__dirname, `__ai_responses__/todo-inspector-element-${repeatIndex}.json`), JSON.stringify(aiResponse, null, 2), { encoding: 'utf-8'});
    expect(filterUnStableinf).toMatchFileSnapshot('./__snapshots__/todo_inspector.test.ts.snap');
  }, {
    timeout: 99999,
  });
});

