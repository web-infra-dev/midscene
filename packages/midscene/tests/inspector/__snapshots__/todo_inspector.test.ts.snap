[
  {
    "elements": [
      {
        "id": "2",
      },
    ],
    "error": [],
    "prompt": "任务输入框",
  },
  {
    "elements": [
      {
        "id": "8",
      },
    ],
    "error": [],
    "prompt": "任务列表中的第二项",
  },
  {
    "elements": [
      {
        "id": "9",
      },
    ],
    "error": [],
    "prompt": "第二项任务右边的删除按钮",
  },
  {
    "elements": [
      {
        "id": "10",
      },
    ],
    "error": [],
    "prompt": "任务列表中第三项左边的勾选按钮",
  },
  {
    "elements": [
      {
        "id": "15",
      },
    ],
    "error": [],
    "prompt": "任务列表下面的 Completed 状态按钮",
  },
]